File: summary/ts/24_segment_tree.ts
class SegmentTree {
    private tree: number[];
    private n: number;

    constructor(data: number[]) {
        this.n = data.length;
        this.tree = new Array(2 * this.n);
        this.build(data);
    }

    private build(data: number[]): void {
        // Initialize leaves
        for (let i = 0; i < this.n; i++) {
            this.tree[this.n + i] = data[i];
        }

        // Build the tree by calculating parents
        for (let i = this.n - 1; i > 0; i--) {
            this.tree[i] = this.tree[2 * i] + this.tree[2 * i + 1];
        }
    }

    public update(index: number, value: number): void {
        // Set value at position p
        index += this.n;
        this.tree[index] = value;

        // Move upward and update parents
        while (index > 1) {
            index = Math.floor(index / 2);
            this.tree[index] = this.tree[2 * index] + this.tree[2 * index + 1];
        }
    }

    public query(left: number, right: number): number {
        // sum on interval [left, right)
        let result = 0;
        left += this.n;
        right += this.n;

        while (left < right) {
            if (left % 2 === 1) {
                result += this.tree[left];
                left++;
            }
            if (right % 2 === 1) {
                right--;
                result += this.tree[right];
            }
            left = Math.floor(left / 2);
            right = Math.floor(right / 2);
        }

        return result;
    }
}

// Example usage
const data = [1, 3, 5, 7, 9, 11];
const segTree = new SegmentTree(data);

// Query the sum in the range [1, 4)
console.log(segTree.query(1, 4)); // 3 + 5 + 7 = 15

// Update the value at index 2 to 10
segTree.update(2, 10);

// Query again
console.log(segTree.query(1, 4)); // 3 + 10 + 7 = 20

File: summary/python/24_segment_tree.py
class SegmentTree:
    def __init__(self, data):
        self.n = len(data)
        self.tree = [0] * (2 * self.n)
        self.build(data)

    def build(self, data):
        # Initialize leaves
        for i in range(self.n):
            self.tree[self.n + i] = data[i]
        
        # Build the tree by calculating parents
        for i in range(self.n - 1, 0, -1):
            self.tree[i] = self.tree[2 * i] + self.tree[2 * i + 1]

    def update(self, index, value):
        # Set value at position index
        index += self.n
        self.tree[index] = value

        # Move upward and update parents
        while index > 1:
            index //= 2
            self.tree[index] = self.tree[2 * index] + self.tree[2 * index + 1]

    def query(self, left, right):
        # Sum on interval [left, right)
        result = 0
        left += self.n
        right += self.n

        while left < right:
            if left % 2 == 1:
                result += self.tree[left]
                left += 1
            if right % 2 == 1:
                right -= 1
                result += self.tree[right]
            left //= 2
            right //= 2

        return result

# 使用例
if __name__ == "__main__":
    data = [1, 3, 5, 7, 9, 11]
    seg_tree = SegmentTree(data)

    # 配列の範囲 [1, 4) の合計をクエリ
    print(seg_tree.query(1, 4))  # 3 + 5 + 7 = 15

    # インデックス 2 の値を 10 に更新
    seg_tree.update(2, 10)

    # 再度クエリ
    print(seg_tree.query(1, 4))  # 3 + 10 + 7 = 20
